### PR TITLE
Fixed some AI exceptions

### PR DESCRIPTION
**Pull request by Grummel7**
This should fix the exceptions mentioned in #3718 

Btw, can anybody tell me, whether old AIFleetMission.py line 608 makes sense:
`if assertion_fails(new_fleet_id and new_fleet_id != INVALID_ID):`
Common sense says checking for INVALID_ID should do.

### CODE REVIEW

**Comment by agrrr3** [comment]
> Btw, can anybody tell me, whether old AIFleetMission.py line 608 makes sense: `if assertion_fails(new_fleet_id and new_fleet_id != INVALID_ID):` Common sense says checking for INVALID_ID should do.

not sure what you mean with common sense here. 
split_ship_from_fleet returns without values in some places, so the check is necessary

**Comment by Grummel7** [comment]
> not sure what you mean with common sense here. split_ship_from_fleet returns without values in some places, so the check is necessary

Well, my "common sense" thinking was that if there is a constant named INVALID_ID (-1), then 0 should be a valid fleet id and the assertion is bad. May not be a issue, I guess a split off fleet will never get id 0, but the code shouldn't rely on such an implementation detail.

Indeed I missed the fact that fleet_id could also be None, but then the check should probably better written as:
`if assertion_fails(new_fleet_id != None and new_fleet_id != INVALID_ID):`

File: default/python/AI/InvasionAI.py [hunk 1]
```diff
@@ -327,14 +327,13 @@ def _get_path_from_capital(planet: "fo.planet") -> Tuple[Sequence[SystemId], int
     """
     universe = fo.getUniverse()
     capitol_id = PlanetUtilsAI.get_capital()
-    if capitol_id:
+    if capitol_id != INVALID_ID:
```
Comment (geoffthemedio) [inline default/python/AI/InvasionAI.py]: don't also need to test for `None`?

**Comment by geoffthemedio** [comment]
>... the check should probably better written as: `if assertion_fails(new_fleet_id != None and new_fleet_id != INVALID_ID):`

I'm informed that `if assertion_fails(new_fleet_id and new_fleet_id != INVALID_ID):` would be the stanard way to write that...

File: default/python/AI/InvasionAI.py (hunk 1 above)
Comment (Grummel7) [reply default/python/AI/InvasionAI.py]: Argh, these checks are a pain in the ass, some functions return None, some INVALID_ID, perhaps some do both. At the moment get_capital() does not return None, but I agree it's safer to check both.

And there is still the question whether 0 is a valid id, because the "standard check" `if id and id != INVALID_ID` also yields false if id is 0

Comment (geoffthemedio) [reply default/python/AI/InvasionAI.py]: There won't normally be a planet with id 0. Might be a system 0, though.

**Comment by geoffthemedio** [comment]
> > ... the check should probably better written as: `if assertion_fails(new_fleet_id != None and new_fleet_id != INVALID_ID):`
> 
> I'm informed that `if assertion_fails(new_fleet_id and new_fleet_id != INVALID_ID):` would be the stanard way to write that...

after further discussion, according to PEP8, it should be `if assertion_fails(new_fleet_id is not None and new_fleet_id != INVALID_ID):`

**Comment by Cjkjvfnby** [comment]
> after further discussion, according to PEP8, it should be if assertion_fails(new_fleet_id is not None and new_fleet_id != INVALID_ID):


I'd say if our linter does not catch it, its OK. 

I'd prefer something like `new_fleet_id  not in (None, INVALID_ID)`

File: default/python/AI/AIFleetMission.py [hunk 2]
```diff
@@ -596,16 +596,19 @@ def issue_fleet_orders(self):
                             )
                             allowance = CombatRatingsAI.rating_needed(fleet_remaining_rating, fleet_min_rating)
                             debug("May release ships with total rating of %.1f", allowance)
-                            ship_ids = list(self.fleet.get_object().shipIDs)
+                            org_fleet = self.fleet.get_object().shipIDs
```
Comment (Cjkjvfnby) [inline default/python/AI/AIFleetMission.py]: `org_fleet` is a bit confusing name, since it collection of ships.  

`for ... in ...` and `list()` uses the same things inside, which implies that we could just skip wrapping to list. 

```python
for ship_id  in self.fleet.get_object().shipIDs:
```

File: default/python/AI/InvasionAI.py [hunk 3]
```diff
@@ -327,14 +327,13 @@ def _get_path_from_capital(planet: "fo.planet") -> Tuple[Sequence[SystemId], int
     """
     universe = fo.getUniverse()
     capitol_id = PlanetUtilsAI.get_capital()
-    if capitol_id:
+    if capitol_id and capitol_id != INVALID_ID:
```
Comment (Cjkjvfnby) [inline default/python/AI/InvasionAI.py]: `PlanetUtilsAI.get_capital()` will return id or INVALID_ID, so we could drop` capitol_id and` part

File: default/python/AI/ProductionAI.py [hunk 4]
```diff
@@ -135,6 +135,7 @@ def generate_production_orders():
     building_ratio = aistate.character.preferred_building_ratio([0.4, 0.35, 0.30])
     capital_id, homeworld, capital_system_id = _get_capital_info()
     current_turn = fo.currentTurn()
+    production_queue = empire.productionQueue
```
Comment (Cjkjvfnby) [inline default/python/AI/ProductionAI.py:138]: I'd put it inside else branch.

**Review by Cjkjvfnby — APPROVED** [review]
Looks good. I left a couple of minor comments.

**Comment by Grummel7** [comment]
> `org_fleet` is a bit confusing name, since it collection of ships.

Perhaps I should simply work without an extra name, it is used only once anyway.

> `for ... in ...` and `list()` uses the same things inside, which implies that we could just skip wrapping to list.
> 
> ```python
> for ship_id  in self.fleet.get_object().shipIDs:
> ```

As far as I understand it, the list() wrapper is a way to create a copy, it was there before. Whether the copy is necessary is beyond my python expertise. The problem is that self.fleet is modified within the loop.

> I'd prefer something like `new_fleet_id not in (None, INVALID_ID)`

I like that one, could use it for capital case below as well.

> `PlanetUtilsAI.get_capital()` will return id or INVALID_ID, so we could drop `capitol_id and` part

Originally I only checked for INVALID_ID, but Geoff complained about it. And I think he's right, in doubt we should rather check both values, since it is less likely to be affected by future changes in other parts of the code.

> I'd put it inside else branch.

That's were it was, but the name is accessed after the if...else, that's what I put it outside.

**Comment by Cjkjvfnby** [comment]
@Grummel7 It will be much easier if you will pull answers in the comment thread.  In that case, I can just resolve them.  Also, they will be close to the code we are talking about.

> As far as I understand it, the list() wrapper is a way to create a copy, it was there before. 
Whether the copy is necessary is beyond my python expertise. The problem is that self.fleet is modified within the loop.

> it was there before

In this codebase, it's not a good argument. For many reasons, a lot of strange things are here. Keeping them does not make code better. 

> The problem is that self.fleet is modified within the loop.

We use `fleet.get_object().shipIDs` which is already a copy. Not critical for that PR

> Originally I only checked for INVALID_ID, but Geoff complained about it. And I think he's right, in doubt we should rather check both values, since it is less likely to be affected by future changes in other parts of the code.

`PlanetUtilsAI.get_capital()` never returns None. And changing it will require checking all usages. I don't expect this change will ever happen, so we could keep things simple for that case. Instead of that other places that return INVALID_ID or None, should be changed to return only INVALID id, but this is out of the scope of this PR.

> That's were it was, but the name is accessed after the if...else, that's what I put it outside.

Now I see it.  Big functions are a bad thing.

File: default/python/AI/InvasionAI.py (hunk 1 above)
Comment (Grummel7) [reply default/python/AI/InvasionAI.py]: Actually yes, there is a system with ID 0. Just had a test game in which one AI started in it. It never left the system. Haven't yet figured out why. Log says:

`21:08:28.839817 {0x00007fa656944780} [warn] python : target.py:19 : Target is invalid inaccessible system_-1<>`
`21:08:28.839890 {0x00007fa656944780} [debug] python : MoveUtilsAI.py:51 : Requesting path for fleet F_3902<Recon Fleet 3902> from inaccessible sys>`
`21:08:28.839993 {0x00007fa656944780} [warn] python : pathfinding.py:204 : Requested path between invalid systems.`

But this is a separate issue...

**Comment by Grummel7** [comment]
> @Grummel7 It will be much easier if you will pull answers in the comment thread. In that case, I can just resolve them. Also, they will be close to the code we are talking about.

Okay, will do next time. There'll surely be bigger PRs

> We use `fleet.get_object().shipIDs` which is already a copy. Not critical for that PR

Actually, if it is a copy, my length check works on a copy, making it pointless.
But I guess you are right, since the object comes straight from C++.
So I simplified it by not using extra names at all.

> `PlanetUtilsAI.get_capital()` never returns None. And changing it will require checking all usages. I don't expect this change will ever happen, so we could keep things simple for that case. Instead of that other places that return INVALID_ID or None, should be changed to return only INVALID id, but this is out of the scope of this PR.

Well, I agree on that as well, so I changed it back.

**Comment by geoffthemedio** [comment]
I'm not clear about the status of this, but @Cjkjvfnby can you merge when it's ready?

**Review by Cjkjvfnby — APPROVED** [review]

**Comment by geoffthemedio** [comment]
I'm getting some AI errors after loading a save, which are plausibly related to these changes...
````
[05 Mar 23:50:56] AI_2: AI_Error: AI script error : AIstate.py:__clean_fleet_roles():938  - Fleet F_186269<Fleet 186269> has no valid system.
[05 Mar 23:50:56] AI_1: AI_Error: AI script error : AIstate.py:__clean_fleet_roles():938  - Fleet F_183699<Battle Fleet 183699> has no valid system.
[05 Mar 23:50:56] AI_3: AI_Error: AI script error : AIstate.py:__clean_fleet_roles():938  - Fleet F_186207<Battle Fleet 186207> has no valid system.
[05 Mar 23:50:56] AI_2: AI_Error: AI script error : AIstate.py:__clean_fleet_roles():938  - Fleet F_186819<Battle Fleet 186819> has no valid system.
[05 Mar 23:50:56] AI_1: AI_Error: AI script error : AIstate.py:__clean_fleet_roles():938  - Fleet F_186764<Battle Fleet 186764> has no valid system.
[05 Mar 23:50:56] AI_3: AI_Error: AI script error : AIstate.py:__clean_fleet_roles():938  - Fleet F_188091<Fleet 188091> has no valid system.
[05 Mar 23:50:56] AI_2: AI_Error: AI script error : AIstate.py:__clean_fleet_roles():938  - Fleet F_189108<Battle Fleet 189108> has no valid system.
[05 Mar 23:50:56] AI_1: AI_Error: AI script error : AIstate.py:__clean_fleet_roles():938  - Fleet F_189053<Battle Fleet 189053> has no valid system.
[05 Mar 23:50:56] AI_3: AI_Error: AI script error : AIstate.py:__clean_fleet_roles():938  - Fleet F_188494<Battle Fleet 188494> has no valid system.
[05 Mar 23:50:56] AI_3: AI_Error: AI script error : AIstate.py:__clean_fleet_roles():938  - Fleet F_188660<Fleet 188660> has no valid system.
[05 Mar 23:50:56] AI_3: AI_Error: AI script error : AIstate.py:__clean_fleet_roles():938  - Fleet F_188671<Fleet 188671> has no valid system.
````

**Comment by Grummel7** [comment]
> I'm getting some AI errors after loading a save, which are plausibly related to these changes...
> 
> ```
> [05 Mar 23:50:56] AI_2: AI_Error: AI script error : AIstate.py:__clean_fleet_roles():938  - Fleet F_186269<Fleet 186269> has no valid system.
> ...
> ```

This happened with current master? I do not see how this PR should affect fleet positions.
Or have you merged #3723 in your environment? That one at least does change code related to fleet positions.

**Comment by geoffthemedio** [comment]
I thought it was master, but I'm not able to reproduce the problem now...
